core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -16,417 +16,444 @@
  * @fileoverview Expression syntax tree service.
  */
 
-require('expressions/expression-parser.service.ts');
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
 
-angular.module('oppia').factory('ExpressionSyntaxTreeService', [
-  'ExpressionParserService', 'PARAMETER_TYPES',
-  function(ExpressionParserService, PARAMETER_TYPES) {
-    // Exceptions that can be thrown from the evaluation of expressions.
-    const ExpressionError = function() {};
-    ExpressionError.prototype = new Error();
-    ExpressionError.prototype.constructor = ExpressionError;
+import { AppConstants } from 'app.constants';
+import { ExpressionParserService } from
+  'expressions/expression-parser.service.ts';
 
-    var ExprUndefinedVarError = function(varname, envs) {
-      this.varname = varname;
-      this.envs = envs;
-    };
-    ExprUndefinedVarError.prototype = new ExpressionError();
-    ExprUndefinedVarError.prototype.constructor = ExprUndefinedVarError;
-    ExprUndefinedVarError.prototype.name = 'ExprUndefinedVarError';
-    ExprUndefinedVarError.prototype.toString = function() {
-      return this.name + ': ' + this.varname + ' not found in ' + this.envs;
-    };
+class ExpressionError extends Error {
+  constructor() {
+    super();
+    // NOTE TO DEVELOPERS: As per recommendation of TypeScript
+    // extending Error would need to adjust the protoype manually
+    // This code below would restore the protoype chain
+    // Source: https://stackoverflow.com/a/58417721/4859885
+    Object.setPrototypeOf(this, new.target.prototype);
+    this.name = new.target.name;
```

Comment:
Could you please add a simple test to confirm these work?

Reply:
@brianrodri Would this be enough or should I take all the expression error classes and place them in an array?
```
 describe('ExpressionError', () => {
    let expressionError: ExpressionError;

    it('should extend Error object', () => {
      expressionError = new ExpressionError();

      expect(expressionError.name).toBe('ExpressionError');
      expect(expressionError instanceof Error).toBe(true);
    });
  });
```

Reply:
Placing them in an array makes most sense, that way you can reach 100% frontend coverage as well.